File: src/common/322/518.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        @cache  # 缓存装饰器，避免重复计算 dfs 的结果（记忆化）
        def dfs(i: int, c: int) -> int:
            if i < 0:
                return 1 if c == 0 else 0
            if c < coins[i]:
                return dfs(i - 1, c)
            return dfs(i - 1, c) + dfs(i, c - coins[i])
        return dfs(len(coins) - 1, amount)



class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        n = len(coins)
        f = [[0] * (amount + 1) for _ in range(n + 1)]
        f[0][0] = 1
        for i, x in enumerate(coins):
            for c in range(amount + 1):
                if c < x:
                    f[i + 1][c] = f[i][c]
                else:
                    f[i + 1][c] = f[i][c] + f[i + 1][c - x]
        return f[n][amount]



class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        f = [1] + [0] * amount
        for x in coins:
            for c in range(x, amount + 1):
                f[c] += f[c - x]
        return f[amount]

File: src/draft/0draft1028/518.js
/**
 * @param {number} amount
 * @param {number[]} coins
 * @return {number}
 */
var change = function (amount, coins) {
  const n = coins.length;
  let memo = Array.from({ length: n }, () => new Array(amount + 1).fill(-1));
  const dfs = (i, c) => {
    // 现在的位置是i 还需要凑足c元
    if (i < 0) {
      return c == 0 ? 1 : 0;
    }
    if (memo[i][c] != -1) return memo[i][c];
    if (c < coins[i]) {
      return (memo[i][c] = dfs(i - 1, c));
    }

    return (memo[i][c] = dfs(i - 1, c) + dfs(i, c - coins[i]));
  };
  return dfs(n - 1, amount);
};

var change = function (amount, coins) {
  const n = coins.length;
  let dp = Array.from({ length: n + 1 }, () => new Array(amount + 1).fill(0));
  dp[0][0] = 1;
  for (let i = 0; i < n; i++) {
    let x = coins[i];
    for (let c = 0; c <= amount; c++) {
      if (c < x) {
        dp[i + 1][c] = dp[i][c];
      } else {
        dp[i + 1][c] = dp[i][c] + dp[i + 1][c - x];
      }
    }
  }
  return dp[n][amount];
};

var change = function (amount, coins) {
  const n = coins.length;
  let dp = new Array(amount + 1).fill(0);
  dp[0] = 1;
  for (let i = 0; i < n; i++) {
    let x = coins[i];
    for (let c = x; c <= amount; c++) {
      dp[c] += dp[c - x];
    }
  }
  return dp[n][amount];
};
